orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,206 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
   graph.AddInitializedTensor(proto_data);
 }
 
+Status FindAllConnectedNodes(Graph& graph,
+                             const Node* node,
+                             std::vector<const Node*>& connected_nodes,
+                             std::set<const NodeArg*>& connected_inputs,
+                             std::set<const NodeArg*>& connected_outputs
+                             ) {
+  ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+      node->InputDefs(),
+      [&](const NodeArg& node_arg, size_t /*index*/) {
+        if (graph.IsInputsIncludingInitializers(&node_arg)) {
+          connected_inputs.insert(&node_arg);
+        } else {
+          const Node* producer_node = graph.GetProducerNode(node_arg.Name());
+          if (producer_node == nullptr) {
+            // got nullptr as producer node. This could be because the input is a constant op which will be optimized
+            // away. Print out this information and continue.
+            LOGS_DEFAULT(WARNING) << "Cannot find producer node for node_arg: " << node_arg.Name() << ". Skipping this node.";
+          } else {
+            connected_nodes.push_back(producer_node);
+          }
+        }
+        return Status::OK();
+      }));
+
+  ORT_THROW_IF_ERROR(node->ForEachWithIndex(
+      node->OutputDefs(),
+      [&](const NodeArg& node_arg, size_t /*index*/) {
+        if (!graph.IsOutput(&node_arg)) {
+          std::vector<const Node*> consumer_nodes = graph.GetConsumerNodes(node_arg.Name());
+          connected_nodes.insert(std::end(connected_nodes), consumer_nodes.begin(), consumer_nodes.end());
+
+        } else {
+          connected_outputs.insert(&node_arg);
+        }
+        return Status::OK();
+      }));
+  return Status::OK();
+}
+
+// PipelineStageNodeGroup groups nodes that share the same input initializer and belong to the same stage.
+// It is used to distinguish other nodes that share the same input initializer but belong to
+// other pipeline partitions after split.
+struct PipelineStageNodeGroup {
+  const size_t stage_id;
+
+  // Vector of nodes that have the same initializer input and belong to the same stage. Noted that
+  // the consumer nodes of a particular initializer can be more than one, so we need a vector to store those
+  // nodes.
+  std::vector<Node*> nodes;
+  PipelineStageNodeGroup(const size_t stage, std::vector<Node*>& node) : stage_id(stage), nodes(std::move(node)){};
```

Comment:
Is the "node" (singular) in the parameters indicating that a vector of length 1 should be supplied, or just to avoid the name clash with the field?  If it's for naming then could we keep a plural and disambiguate some other way.  If the vector is of length 1 then can we assert that.

Reply:
Yeah it should be plural as it is a group of nodes. Fixed. Thanks.

---
In reply to: [452259345](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452259345) [](ancestors = 452259345)